provider/oci/common/client.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package common
+
+import (
+	"context"
+	"crypto/rsa"
+	"fmt"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/instance"
+	"github.com/juju/juju/network"
+
+	ociCommon "github.com/oracle/oci-go-sdk/common"
+	ociCore "github.com/oracle/oci-go-sdk/core"
+	ociIdentity "github.com/oracle/oci-go-sdk/identity"
+)
+
+type jujuConfigProvider struct {
+	key            []byte
+	keyFingerprint string
+	passphrase     string
+	tenancyOCID    string
+	userOCID       string
+	region         string
+}
+
+type ociClient struct {
+	ociCore.ComputeClient
+	ociCore.BlockstorageClient
+	ociCore.VirtualNetworkClient
+	ociIdentity.IdentityClient
+
+	ociCommon.ConfigurationProvider
+}
+
+// NewJujuConfigProvider returns a new ociCommon.ConfigurationProvider instance
+func NewJujuConfigProvider(user, tenant string, key []byte, fingerprint, passphrase, region string) ociCommon.ConfigurationProvider {
```

Comment:
This function signature is already getting a little bit unweildy. Usually its better to have an exposed Struct once the number of parameters is large enough.

What do you think about changing this to 
type JujuConfigProvider struct {
  Key string
  KeyFingerprint string
...
}

And then we would add a Validate() sort of function that would be called at the beginning of NewOCIClient?

Generally, having a constructor that takes 10 anonymous attributes is worse than exposing a Parameter struct type because then you get to see the Key: Value pairs.

We'd like to see some sort of Validate() function anyway, so that you get an early "this doesn't look like valid configuration" rather than having it fail deep in some function much later.